File: src/com/train/algorithm/list/implementInJS/Subsets.js
// 子集问题
// leetcode-78
var subsets = function(nums) {
	const temp = [];
	const res = [];
	const len = nums.length;
	const dfs = (cur) => {
		if(cur === len){
			res.push(temp.slice());
			return;
		}
		temp.push(nums[cur]);
		dfs(cur + 1,nums);
		temp.pop(temp.length - 1);
		dfs(cur + 1,nums);
	}
	dfs(0,nums);
	return res;
};

//leetcode-46
var permute = function(nums) {
	const temp = [],res = [];
	const len = nums.length;
	dfs(nums,len,[]);
	return res;


	function dfs(n,k,used){
		if(temp.length === k){
			res.push(temp.slice());
			return;
		}
		for(let i = 0;i < k;i++){
			if(used[i]){
				continue;
			}
			temp.push(n[i]);
			used[i] = true;
			dfs(n,k,used);
			temp.pop();
			used[i] = false;
		}
	}
			
};

// leetcode-784
var letterCasePermutation = function(s) {
	const ans = [], n = s.length;   
	function dfs(index, tmp) {
		if (index === n) return ans.push(tmp);
		if (isNaN(s[index])) {
			dfs(index + 1, tmp + s[index].toLowerCase());
			dfs(index + 1, tmp + s[index].toUpperCase());
		} else {
			dfs(index + 1, tmp + s[index]);
		}
	}
	dfs(0, '');
	return ans;
};

File: src/com/train/algorithm/search/implementInPython/Subsets.py
class Solution(object):
    def dfs(self, nums, path, res):
        res.append(path)
        for i in range(len(nums)):
            self.dfs(nums[i+1:], path + [nums[i]], res)
    def subsets(self, nums):
        res = []
        self.dfs(nums, [], res)
        return res


        # def backtracking(start, path):
        #
        #     res.append(path[:])
        #
        #     for i in range(start, len(nums)):
        #
        #         path.append(nums[i])
        #
        #         backtracking(i+1, path)
        #
        #         path.pop()
        # res = []
        # backtracking(0, [])
        # return res